File: stud/zhilin/Lab1/task_1.4/main.cpp
#include <bits/stdc++.h>

using namespace std;
using matrix = vector<vector<double> >;


void print_matrix(const matrix& matrix1) {
    for(const auto& vect: matrix1) {
        for (auto x: vect)
            cout << x << " ";
        cout << endl;
    }
}


pair<int, int> get_indexes(const matrix& matrix1){
    double k = matrix1[0][1];
    pair<int, int> indexes = make_pair(0, 1);
    int n = matrix1.size(), m = matrix1[0].size();
    for (int i = 0; i < n; i++)
        for (int j = i+1; j < m; j++)
            if (abs(matrix1[i][j]) > k) {
                k = abs(matrix1[i][j]);
                indexes = make_pair(i, j);
            }
    return indexes;
}


double get_phi(double a_ij, double a_ii, double a_jj){
    if (a_ii == a_jj)
        return 3,1415926535/4;
    return atan(2*a_ij/(a_ii-a_jj)) / 2;
}


matrix multiple_matrix(const matrix& matrix1, const matrix& matrix2) {
    int n1 = matrix1.size(), m1 = matrix1[0].size(), m2 = matrix2[0].size();
    matrix res(n1);
    for (int i=0; i<n1; i++)
        for (int j=0; j<m2; j++)
            res[i].push_back(0);

    for (int i=0; i<n1; i++) {
        for (int j=0; j<m2; j++) {
            double cntr = 0;
            for (int k=0; k<m1; k++)
               cntr += matrix1[i][k] * matrix2[k][j];
            res[i][j] = cntr;
        }
    }
    return res;
}


matrix get_E_matrix(int n){
    matrix E(n, vector<double>(n, 0));
    for(int i=0; i<n; i++)
        E[i][i] = 1;
    return E;
}


matrix get_U_matrix(const matrix& matrix1){
    matrix res = get_E_matrix(matrix1.size());
    int i_max, j_max;
    tie(i_max, j_max) = get_indexes(matrix1);
    double phi = get_phi(matrix1[i_max][j_max], matrix1[i_max][i_max], matrix1[j_max][j_max]);
    res[i_max][i_max] = cos(phi);
    res[j_max][j_max] = cos(phi);
    res[i_max][j_max] = -sin(phi);
    res[j_max][i_max] = sin(phi);
    return res;
}


double get_eps(const matrix& matrix1){
    double eps = 0;
    int n=matrix1.size(), m=matrix1[0].size();
    for (int i=0; i<n; i++)
        for (int j=i+1; j<m; j++)
            eps += matrix1[i][j]*matrix1[i][j];
    return sqrt(eps);
}


matrix transposed(const matrix& matrix1){
    int n = matrix1.size(), m = matrix1[0].size();
    matrix res(m, vector<double>(n));
    for (int i=0; i<n; i++)
        for (int j=0; j<m; j++)
            res[j][i] = matrix1[i][j];
    return res;
}


pair<vector<double>, matrix> Jacobi_method(matrix& coeff_matrix, double EPS){
    int n = coeff_matrix.size();
    matrix eigenvectors = get_E_matrix(n);
    while (get_eps(coeff_matrix) > EPS){
        matrix U = get_U_matrix(coeff_matrix);
        eigenvectors = multiple_matrix(eigenvectors, U);
        matrix transposed_U = transposed(U);
        coeff_matrix = multiple_matrix(multiple_matrix(transposed_U, coeff_matrix), U);
    }

    vector<double> eigenvalues(n);
    for (int i=0; i<n; i++)
        eigenvalues[i] = coeff_matrix[i][i];
    return make_pair(eigenvalues, eigenvectors);
}


int main() {
    matrix coefficient_matrix{
        {4, 7, -1},
        {7, -9, -6},
        {-1, -6, -4}
    };


    vector<double> eigenvalues;
    matrix eigenvectors;
    tie(eigenvalues, eigenvectors) = Jacobi_method(coefficient_matrix, 0.01);
    int n = eigenvalues.size();

    cout << "Eigenvalues:" << endl;
    for (int i=0; i<n; i++)
        cout << "lambda_" << i+1 << " = " << eigenvalues[i] << endl;
    cout << endl;

    cout << "Eigenvectors:" << endl;
    for (int i=0; i<n; i++){
        cout << "x_" << i+1 << " = ";
        for(double element: eigenvectors[i])
            cout << element << " ";
        cout << endl;
    }

    return 0;
}

File: stud/zhilin/Lab1/task_1.4/main.py
import math

from matrix import *
from matrix import Vector, Matrix


def get_indexes(matrix1: Matrix) -> tuple[int, int]:
    """
    return indexes of maximum non-diagonal element
    """
    m: float = matrix1[0][1]
    indexes: tuple[int, int] = (0, 1)
    for i in range(matrix1.n):
        for j in range(i+1, matrix1.m):
            if abs(matrix1[i][j]) > m:
                m = abs(matrix1[i][j])
                indexes = (i, j)
    return indexes


def get_phi(a_ij: float, a_ii: float, a_jj: float) -> float:
    if a_ii == a_jj:
        return math.pi/4
    return math.atan(2*a_ij/(a_ii-a_jj)) / 2


def get_U_matrix(matrix1: Matrix) -> Matrix:
    res = [[1 if i == j else 0 for j in range(matrix1.n)] for i in range(matrix1.n)]
    i_max, j_max = get_indexes(matrix1)
    phi = get_phi(a_ij=matrix1[i_max][j_max], a_ii=matrix1[i_max][i_max], a_jj=matrix1[j_max][j_max])
    res[i_max][i_max] = math.cos(phi)
    res[j_max][j_max] = math.cos(phi)
    res[i_max][j_max] = -math.sin(phi)
    res[j_max][i_max] = math.sin(phi)
    return Matrix(res)


def get_eps(matrix1: Matrix) -> float:
    eps = 0
    for i in range(matrix1.n):
        for j in range(i+1, matrix1.m):
            eps += matrix1[i][j]**2
    return math.sqrt(eps)


def Jacobi_method(coeff_matrix: Matrix, EPS: float) -> tuple[Vector, Matrix]:
    eigenvectors = Matrix([[1 if i == j else 0 for j in range(coeff_matrix.n)]for i in range(coeff_matrix.n)])
    while get_eps(coeff_matrix) > EPS:
        U = get_U_matrix(coeff_matrix)
        eigenvectors = multiple_matrix(eigenvectors, U)
        coeff_matrix = multiple_matrix(multiple_matrix(U.transposed(), coeff_matrix), U)
    eigenvalues = Vector([coeff_matrix[i][i] for i in range(coeff_matrix.n)])
    return eigenvalues, eigenvectors


if __name__ == "__main__":
    coefficient_matrix = Matrix([
        [4, 7, -1],
        [7, -9, -6],
        [-1, -6, -4]
    ])

    values, vectors = Jacobi_method(coefficient_matrix, EPS=0.01)
    print("Собственные значения:", *[f'λ_{i+1} = {values[i]}' for i in range(values.n)], sep='\n', end='\n\n')

    print("Собственные векторы:")
    for i in range(vectors.n):
        print(f'x_{i+1} = {[vectors[j][i] for j in range(vectors.n)]}')
